src/software/geom/util.cpp

```diff
@@ -1251,6 +1116,58 @@ std::optional<Segment> mergeFullyOverlappingSegments(Segment segment1, Segment s
     }
 }
 
+std::vector<Segment> reduceParallelSegments(std::vector<Segment> segments)
+{
+    // If there are no segments, return nullopt
+    if (segments.size() == 0)
+    {
+        return std::vector<Segment>();
+    }
+    else if (segments.size() == 1)
+    {
+        // If there is only 1 segments, it is unique and should be returned
+        return segments;
+    }
+
+    // Check that all segments are collinear
+    for (unsigned int i = 0; i < segments.size() - 1; i++)
+    {
+        if (!collinear(segments[i], segments[i + 1]))
+        {
+            return std::vector<Segment>();
```

Comment:
I'm confused what this means. I would expect that if we gave this function 4 segments, 3 of which are parallel and a forth which is not, it would reduce all the parallel ones and leave the non-parallel ones. But if I understand this correctly, in this case we'd just return an empty vector?

Reply:
I think I'd prefer that either:
- we change this to match the behavior I was expecting
- we throw an exception here

@MathewMacDougall your thoughts?

Reply:
I agree this function is confusing. My assumption of what it does based on the name doesn't match the implementation.

A few suggestions:
* Either of what gareth suggested.
* Change the name of the function. `reduce` sounds funny to me, this really seems more like `mergeParallelSegments` or something like that.
* Maybe also provide a line for all segments to be projected onto before being merged? Then you can guarantee that you will never have non-parallel lines. Something like `projectAndMergeSegments(Line, Segments)`. Then we wouldn't need to worry about exceptions and stuff like that.

Reply:
Yeah I get what you're saying here.

I think making the function behave as your expected makes sense from a general use scenario.
I wrote is like this because I intended it to be used in the shot Eval stuff I wrote which will not work with non-parallel segments

Reply:
That's fine, but `reduce` to me doesn't imply that "nothing" will be returned if a non-parallel segment is given. That seems more suited to `isAllSegmentsParallel` or more of a "checking" function. That's what I mean by the name not matching the implementation, not necessarily the fact that the use case is more specific.

Reply:
I'll change the function to what Gareth suggested. 

Parallel segments will be merged, but non-parallel will be left unchanged.

Reply:
I'm hesitant to change the function to merge parallel segments and leave the non-parallel ones to remain in the vector.

Just because every Segment in the vector must be checked for other parallel Segments and complicated the function a lot more from it's intended purpose

Reply:
What about projecting then?

Reply:
Just to be clear, the behavior I was expecting was something like:
```
std::sort(segments.begin(), segments.end(), [](Segment s1, Segment s2){
            return s1.orientation() < s2.orientation
        });

std::accumulate(segments.begin(), segments.end(), std::vector<Segment>(), 
        [](std::vector<Segment> reduced_segments, Segment s){
            if (reduced_segments.size() == 0){
                reduced_segments.emplace_back(s);
            } else if (colinear(reduced_segments.back(), s)){
                reduced_segments.back() = merge(reduced_segments.back(), s);
            } else {
                reduced_segments.emplace_back(s);
            }
        });
```
(somewhat pseudo-code). Basically we sort the lines so that parallel ones will always be adjacent, then perform a reduction where we go through the list and merge parallel segments.

Reply:
For this function I am thinking of changing the parameters so something like:

function(vector<Segment> segs, Vector direction)

That does the same thing as the existing function, but just uses he projection of segs onto the direction Vector

Reply:
Sounds good to me

Reply:
reduceParallelSegments() is deleted and replaced with combineToParallelSegments()

Reply:
This seems OK to me now. @MathewMacDougall - please resolve if you're good with this.